WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt

```diff
@@ -3,25 +3,78 @@ package org.wordpress.android.viewmodel.pages
 import android.arch.lifecycle.MutableLiveData
 import android.arch.lifecycle.ViewModel
+import kotlinx.coroutines.experimental.CommonPool
+import kotlinx.coroutines.experimental.launch
 import org.wordpress.android.R.string
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.models.pages.PageModel
+import org.wordpress.android.networking.PageStore
 import org.wordpress.android.ui.pages.PageItem
 import org.wordpress.android.ui.pages.PageItem.Action
 import org.wordpress.android.ui.pages.PageItem.Action.PUBLISH_NOW
 import org.wordpress.android.ui.pages.PageItem.Action.VIEW_PAGE
 import org.wordpress.android.ui.pages.PageItem.Divider
 import org.wordpress.android.ui.pages.PageItem.Empty
 import org.wordpress.android.ui.pages.PageItem.Page
+import org.wordpress.android.util.AppLog
+import org.wordpress.android.viewmodel.SingleLiveEvent
+import org.wordpress.android.viewmodel.pages.PageListViewModel.PageListState
+import org.wordpress.android.viewmodel.pages.PageListViewModel.PageListState.CAN_LOAD_MORE
+import org.wordpress.android.viewmodel.pages.PageListViewModel.PageListState.DONE
+import org.wordpress.android.viewmodel.pages.PageListViewModel.PageListState.ERROR
+import org.wordpress.android.viewmodel.pages.PageListViewModel.PageListState.FETCHING
+import org.wordpress.android.viewmodel.pages.PageListViewModel.PageListState.LOADING_MORE
 import javax.inject.Inject
 
 class PagesViewModel
-@Inject constructor() : ViewModel() {
-    private val mutableSearchExpanded: MutableLiveData<Boolean> = MutableLiveData()
-    private val mutableSearchResult: MutableLiveData<List<PageItem>> = MutableLiveData()
-    val searchExpanded: LiveData<Boolean> = mutableSearchExpanded
-    val searchResult: LiveData<List<PageItem>> = mutableSearchResult
-
-    fun start() {
-        clear()
+@Inject constructor(private val pageStore: PageStore) : ViewModel() {
+    private val _isSearchExpanded = SingleLiveEvent<Boolean>()
+    val isSearchExpanded: LiveData<Boolean> = _isSearchExpanded
+
+    private val _searchResult: MutableLiveData<List<PageItem>> = MutableLiveData()
+    val searchResult: LiveData<List<PageItem>> = _searchResult
+
+    private val _listState = MutableLiveData<PageListState>()
+    val listState: LiveData<PageListState>
+        get() = _listState
+
+    private val _refreshPages = SingleLiveEvent<Unit>()
+    val refreshPages: LiveData<Unit>
+        get() = _refreshPages
+
+    private var _pages: List<PageModel> = emptyList()
+    val pages: List<PageModel>
+        get() = _pages
+
+    private lateinit var site: SiteModel
+
+    fun start(site: SiteModel) {
+        this.site = site
+
+        clearSearch()
+        reloadPagesAsync()
+    }
+
+    private fun reloadPagesAsync() = launch(CommonPool) {
+        _pages = pageStore.loadPagesFromDb(site)
+        refreshPagesAsync()
+    }
+
+    private fun refreshPagesAsync(isLoadingMore: Boolean = false) = launch(CommonPool) {
```

Comment:
what about making this a suspend function, in the `reloadPagesAsync` it can be called anyway (and probably makes little more sense) and in the `refresh` fun you could just call directly `launch { refreshPages() }`. Not a big deal but it might make it little more understandable. What happens, when you call launch inside of launch? Does it run on a different thread?

Reply:
You can launch new coroutines inside coroutines, that's 100% allowed. What thread it launches on depends on a coroutine context that's specified (or not). If it's not specified it runs on the thread pool. You can read more about this [here](https://github.com/Kotlin/kotlinx.coroutines/blob/master/coroutines-guide.md#coroutine-context-and-dispatchers).

You're probably right about the `reloadPages()`. By making it a suspended function, we can leave it up to the caller what type of coroutine they want to run it on.

Reply:
thanks for the link, I'm still thinking in the Thread world. With coroutines there is really nothing like jumping between threads